src/program.cpp

```diff
@@ -962,142 +987,126 @@ cl_build_status cvk_program::compile_source(const cvk_device* device) {
     {
         static std::mutex clspv_compile_mutex;
         std::lock_guard<std::mutex> clspv_compile_lock(clspv_compile_mutex);
-        status = clspv::CompileFromSourceString(
-            m_source, "", build_options, m_binary.raw_binary(), &m_build_log);
-    }
-    if (status != 0) {
-        cvk_error_fn("failed to compile the program");
-        return CL_BUILD_ERROR;
+        std::vector<std::string> programs;
+        if (m_operation == build_operation::link) {
+            for (auto input_program : m_input_programs) {
+                if (input_program->m_binary_type !=
+                        CL_PROGRAM_BINARY_TYPE_COMPILED_OBJECT &&
+                    input_program->m_binary_type !=
+                        CL_PROGRAM_BINARY_TYPE_LIBRARY) {
+                    return CL_BUILD_ERROR;
+                }
+                programs.push_back(input_program->m_ir);
+            }
+        } else {
+            if (m_source.empty() && !m_ir.empty()) {
+                programs.push_back(m_ir);
+            } else {
+                programs.push_back(m_source);
+            }
+        }
+        m_build_log.clear();
+        if (build_to_ir) {
+            std::vector<uint32_t> ir;
```

Comment:
Looking at the whole chain from clspv to here, there are lots of conversions between uin32_t, steams, strings, etc. We should look at streamlining this a bit more.